File: apps/node-api/src/routes/analyze.ts
/**
 * Analysis submission endpoints. Mirrors the Python API surface:
 *   POST /analyze-snippet, /analyze-file, /analyze-folder,
 *        /analyze-repository (async), /analyze-zip (multipart, async), /reanalyze
 */

import fs from "node:fs";

import {
  Engine,
  DISCLAIMER,
  loadSnippet,
  loadFiles,
  type AnalysisResult,
} from "@aicodecheck/detector-core";
import { Router } from "express";
import multer from "multer";
import { z } from "zod";

import { getSettings } from "../config.js";
import { computeCacheKey, getStore } from "../db.js";
import { enqueueRepository, enqueueZip } from "../queue.js";
import {
  HttpError,
  encryptToken,
  rateLimit,
  validateRepoReference,
  validateSnippet,
  validateUploadSize,
} from "../security.js";
import { asyncHandler, newId } from "./helpers.js";

export const router: Router = Router();
router.use(rateLimit);

const engine = new Engine();

function withDisclaimer(result: AnalysisResult): AnalysisResult {
  if (!result.disclaimer) result.disclaimer = DISCLAIMER;
  return result;
}

// --------------------------------------------------------------------------- //

const snippetSchema = z.object({
  code: z.string(),
  filename: z.string().default("snippet.txt"),
  language: z.string().nullish(),
});

router.post(
  "/analyze-snippet",
  asyncHandler(async (req, res) => {
    const body = snippetSchema.parse(req.body ?? {});
    validateSnippet(body.code);
    const scan = loadSnippet(body.code, {
      filename: body.filename,
      language: body.language ?? null,
    });
    const result = withDisclaimer(engine.analyze(scan, newId()));
    const store = await getStore();
    await store.saveAnalysis(result);
    res.json(result);
  }),
);

const fileSchema = z.object({
  filename: z.string(),
  content: z.string(),
  language: z.string().nullish(),
});

router.post(
  "/analyze-file",
  asyncHandler(async (req, res) => {
    const body = fileSchema.parse(req.body ?? {});
    validateSnippet(body.content);
    const scan = loadSnippet(body.content, {
      filename: body.filename,
      language: body.language ?? null,
    });
    scan.kind = "file";
    const result = withDisclaimer(engine.analyze(scan, newId()));
    const store = await getStore();
    await store.saveAnalysis(result);
    res.json(result);
  }),
);

const folderSchema = z.object({
  files: z.record(z.string()),
  name: z.string().default("uploaded-folder"),
  config: z.unknown().optional(),
});

router.post(
  "/analyze-folder",
  asyncHandler(async (req, res) => {
    const body = folderSchema.parse(req.body ?? {});
    const total = Object.values(body.files).reduce(
      (a, c) => a + Buffer.byteLength(c, "utf8"),
      0,
    );
    validateUploadSize(total);
    const scan = loadFiles(body.files);
    scan.name = body.name;
    const result = withDisclaimer(engine.analyze(scan, newId()));
    const store = await getStore();
    await store.saveAnalysis(result);
    res.json(result);
  }),
);

const repoSchema = z.object({
  repository: z.string(),
  token: z.string().nullish(),
  ref: z.string().nullish(),
  config: z
    .object({ max_files: z.number().int().positive().optional() })
    .partial()
    .optional(),
  async_job: z.boolean().default(true),
});

router.post(
  "/analyze-repository",
  asyncHandler(async (req, res) => {
    const body = repoSchema.parse(req.body ?? {});
    const ref = await validateRepoReference(body.repository);
    const settings = getSettings();
    const maxFiles = body.config?.max_files ?? settings.maxRepoFiles;
    const configRepr = JSON.stringify(body.config ?? {});
    const cacheKey = computeCacheKey(
      "repository",
      `${body.repository}@${body.ref ?? ""}`,
      configRepr,
    );

    const store = await getStore();

    // Serve a cached analysis when one exists and no token is involved.
    if (!body.token) {
      const cached = await store.findCached(cacheKey, settings.analysisCacheTtlSeconds);
      if (cached) {
        res.json({
          job_id: cached.id,
          status: "finished",
          message: "Returned cached analysis.",
          disclaimer: DISCLAIMER,
        });
        return;
      }
    }

    const jobId = newId();
    await store.createJob({
      id: jobId,
      kind: "repository",
      target: body.repository,
      cache_key: cacheKey,
    });
    const tokenEncrypted = body.token ? encryptToken(body.token) : null;
    await enqueueRepository({
      jobId,
      reference: body.repository,
      url: ref.url,
      tokenEncrypted,
      ref: body.ref ?? null,
      maxFiles,
      cacheKey,
    });
    res.json({
      job_id: jobId,
      status: "queued",
      message: "Analysis queued. Poll GET /jobs/{job_id} or GET /analysis/{job_id}.",
      disclaimer: DISCLAIMER,
    });
  }),
);

// Zip upload — stream to a temp file with a running size guard.
const upload = multer({
  storage: multer.diskStorage({
    destination: (_req, _file, cb) => {
      const settings = getSettings();
      fs.mkdirSync(settings.workDir, { recursive: true });
      cb(null, settings.workDir);
    },
    filename: (_req, _file, cb) => cb(null, `upload-${newId()}.zip`),
  }),
  limits: { fileSize: getSettings().maxUploadBytes },
});

router.post(
  "/analyze-zip",
  (req, res, next) => {
    upload.single("file")(req, res, (err: unknown) => {
      if (err) {
        const e = err as { code?: string; message?: string };
        if (e.code === "LIMIT_FILE_SIZE") {
          next(new HttpError(413, `Upload exceeds limit (${getSettings().maxUploadBytes} bytes).`));
        } else {
          next(new HttpError(400, `Upload failed: ${e.message ?? "unknown"}`));
        }
        return;
      }
      next();
    });
  },
  asyncHandler(async (req, res) => {
    const file = (req as unknown as { file?: Express.Multer.File }).file;
    if (!file) throw new HttpError(400, "No 'file' field provided.");
    const store = await getStore();
    const jobId = newId();
    const cacheKey = computeCacheKey(
      "zip",
      `${file.originalname}:${file.size}`,
      "{}",
    );
    await store.createJob({
      id: jobId,
      kind: "zip",
      target: file.originalname || "upload.zip",
      cache_key: cacheKey,
    });
    await enqueueZip({ jobId, zipPath: file.path, cacheKey });
    res.json({
      job_id: jobId,
      status: "queued",
      message: "Analysis queued. Poll GET /jobs/{job_id} or GET /analysis/{job_id}.",
      disclaimer: DISCLAIMER,
    });
  }),
);

const reanalyzeSchema = z.object({ analysis_id: z.string() });

router.post(
  "/reanalyze",
  asyncHandler(async (req, res) => {
    const body = reanalyzeSchema.parse(req.body ?? {});
    const store = await getStore();
    const prior = await store.getAnalysis(body.analysis_id);
    if (!prior) throw new HttpError(404, "Analysis not found.");
    // Inputs (source text) are not retained after analysis for privacy and
    // storage reasons, so re-analysis requires re-submitting the original
    // input. This mirrors the Python API exactly.
    throw new HttpError(
      409,
      "Re-analysis requires re-submitting the original input. Inputs are not " +
        "retained after analysis for privacy and storage reasons.",
    );
    void res;
  }),
);

File: apps/python-api/app/routers/analyze.py
"""Analysis submission endpoints."""

from __future__ import annotations

import os
import tempfile
import uuid

from aiprojectdetector import Engine
from aiprojectdetector.ingestion.loader import load_files, load_snippet
from aiprojectdetector.models import AnalysisResult
from fastapi import APIRouter, Depends, File, Request, UploadFile
from sqlalchemy.orm import Session

from .. import jobs, tasks
from ..config import get_settings
from ..db import get_session
from ..schemas import (
    AnalyzeFileRequest,
    AnalyzeFolderRequest,
    AnalyzeRepositoryRequest,
    AnalyzeSnippetRequest,
    JobCreatedResponse,
    ReanalyzeRequest,
)
from ..security import (
    encrypt_token,
    rate_limit,
    validate_repo_reference,
    validate_snippet,
    validate_upload_size,
)
from ..storage import (
    compute_cache_key,
    create_job,
    find_cached,
    get_analysis,
    save_analysis,
)

router = APIRouter(tags=["analyze"], dependencies=[Depends(rate_limit)])

_engine = Engine()


def _new_id() -> str:
    return uuid.uuid4().hex


@router.post("/analyze-snippet", response_model=AnalysisResult)
def analyze_snippet_endpoint(
    req: AnalyzeSnippetRequest, session: Session = Depends(get_session)
) -> AnalysisResult:
    validate_snippet(req.code)
    scan = load_snippet(req.code, filename=req.filename, language=req.language)
    result = _engine.analyze(scan, analysis_id=_new_id())
    save_analysis(session, result)
    return result


@router.post("/analyze-file", response_model=AnalysisResult)
def analyze_file_endpoint(
    req: AnalyzeFileRequest, session: Session = Depends(get_session)
) -> AnalysisResult:
    validate_snippet(req.content)
    scan = load_snippet(req.content, filename=req.filename, language=req.language)
    scan.kind = "file"
    result = _engine.analyze(scan, analysis_id=_new_id())
    save_analysis(session, result)
    return result


@router.post("/analyze-folder", response_model=AnalysisResult)
def analyze_folder_endpoint(
    req: AnalyzeFolderRequest, session: Session = Depends(get_session)
) -> AnalysisResult:
    total = sum(len(c.encode("utf-8", errors="ignore")) for c in req.files.values())
    validate_upload_size(total)
    scan = load_files(req.files)
    scan.name = req.name
    result = _engine.analyze(scan, analysis_id=_new_id())
    save_analysis(session, result)
    return result


@router.post("/analyze-repository", response_model=JobCreatedResponse)
def analyze_repository_endpoint(
    req: AnalyzeRepositoryRequest, session: Session = Depends(get_session)
) -> JobCreatedResponse:
    validate_repo_reference(req.repository)
    cache_key = compute_cache_key("repository", f"{req.repository}@{req.ref}", req.config.model_dump_json())

    # Serve a cached analysis when one exists and no token is involved.
    if not req.token:
        cached = find_cached(session, cache_key)
        if cached:
            return JobCreatedResponse(
                job_id=cached.id, status="finished", message="Returned cached analysis."
            )

    job_id = _new_id()
    create_job(session, job_id=job_id, kind="repository", target=req.repository, cache_key=cache_key)
    token_enc = encrypt_token(req.token) if req.token else None
    config = req.config.to_options().__dict__

    jobs.enqueue(
        tasks.analyze_repository_task, job_id, req.repository, token_enc, config, cache_key
    )
    return JobCreatedResponse(job_id=job_id, status="queued")


@router.post("/analyze-zip", response_model=JobCreatedResponse)
async def analyze_zip_endpoint(
    request: Request,
    file: UploadFile = File(...),
    session: Session = Depends(get_session),
) -> JobCreatedResponse:
    settings = get_settings()
    os.makedirs(settings.work_dir, exist_ok=True)
    fd, path = tempfile.mkstemp(suffix=".zip", dir=settings.work_dir)
    size = 0
    with os.fdopen(fd, "wb") as out:
        while True:
            chunk = await file.read(1024 * 256)
            if not chunk:
                break
            size += len(chunk)
            if size > settings.max_upload_bytes:
                out.close()
                os.remove(path)
                validate_upload_size(size)  # raises 413
            out.write(chunk)

    job_id = _new_id()
    cache_key = compute_cache_key("zip", f"{file.filename}:{size}", "{}")
    create_job(session, job_id=job_id, kind="zip", target=file.filename or "upload.zip", cache_key=cache_key)
    jobs.enqueue(tasks.analyze_zip_task, job_id, path, {}, cache_key)
    return JobCreatedResponse(job_id=job_id, status="queued")


@router.post("/reanalyze", response_model=AnalysisResult)
def reanalyze_endpoint(
    req: ReanalyzeRequest, session: Session = Depends(get_session)
) -> AnalysisResult:
    """Re-run analysis for a previously analyzed snippet/file/folder.

    Repository/zip re-analysis is best done by re-submitting (their inputs are
    not retained for privacy). For snippets/files/folders the original report's
    target is re-scored with the current engine version.
    """
    from fastapi import HTTPException

    prior = get_analysis(session, req.analysis_id)
    if not prior:
        raise HTTPException(status_code=404, detail="Analysis not found.")
    raise HTTPException(
        status_code=409,
        detail="Re-analysis requires re-submitting the original input. Inputs are not "
        "retained after analysis for privacy and storage reasons.",
    )
